### DISCUSSION TITLE
In JSVM disable template cache while in app.IsDev()

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#5805] [Q&A]
In the past I've found that restarting the JSVM every time that a template is changed can massively slow down the speed of development, especially while on a Raspberry Pi 4.  I got around this by putting the views outside of the `__hooks` directory and then doing a `require` on my own template module.  This approach also required me to add `templateRegistry` to the VM in order to get a fresh one.  The downside is there being just too much setup and customization (as well as other downsides elaborated on below).

**template.js**

```javascript
module.exports = (...frags) => {
  const tplDir = `${__hooks}/../views`
  const paths = frags.map(frag => `${tplDir}/${frag}.html`)

  return getRegistry()
    .loadFiles(...paths)
}

function getRegistry() {
  if ($app.isDev()) {
    return templateRegistry()
  }
  
  return $template
}
```

**A route**

```javascript
routerAdd("get", "/hello/{name}", (e) => {
  const name = e.request.pathValue("name")

  const html = require(`${__hooks}/template.js`)(
    `layouts/layout`,
    `hello`,
  ).render({
    name,
  })

  return e.html(200, html)
})
```

Another downside of this approach is that while not `isDev()` then the views won't be reloaded if they change in production.  This would make "production" tinkering impossible in situations where it may be impossible to fully test the app such as when it's connected to real hardware peripherals.  This takes away the PHP-like live hackability of it.

So all that to say that I think that the template registry needs to be able to not actually store anything while in development mode, or if it's especially requested of it.

There is also the matter of being able to omit watching certain directories (for restarting) while in `isDev` mode.

I'll say that while it might seem minor I've found this to be the largest problem so far for me while developing on low powered hardware.  All other efficiencies are ruined by how templates are handled in that way.

[edit] I'll add once again that this requirement is just for a hobby project, so it's not urgent.

**Comment by chnfyi** [comment]
An idea that **I haven't tested**. 

The workaround is still not absolutely convenient but could solve some of the downsides you mentioned.
- Avoid JSVM restart by moving your views outside pb_hooks, for example in a _templates_ directory
- Have a _templates_ subdirectory that you rename (_e.g. v1, v2, ..._) each time you want to get newly parsed templates

**In JSVM**

```javascript
const dir = $app.isDev() ? 'templates/*/'  : `${__hooks}/views/`

const filenames = ["layout.html", "hello.html"].map((filename) => $filepath.glob(`${dir + filename}`)[0])

const html = $template.loadFiles(
    filenames.join()
).render({
    "name": name,
})
``` 

Then in order to get your views reloaded on change in production you would have to set a GitHub Action to move your files from 'templates/*' directory (_dev structure_) to the 'pb_hooks/views' directory (_production structure_).
